src/stim/circuit/gate_data.h

```diff
@@ -76,8 +115,161 @@ inline uint8_t gate_name_to_id(const char *v, size_t n) {
     return result;
 }
 
+inline uint8_t gate_hash_to_id(uint8_t hash) noexcept {
+    switch (hash) {
```

Comment:
*Definitely* don't construct the table this way! What if I changed the hash function?? Precompute the hash table entries by iterating over the gates and their aliases computing the hash.

I think in C++17 it might be possible to make the old `gate_name_to_id` a constexpr, in which case you could switch over it I think. At the very least this method needs to be saying `case gate_name_to_hash_id("H")` instead of `case 56`.

Reply:
Agree, the big switch is not good.

I've proposed another solution where the parsing is compile time capable via constexpr and type_traits. And I've completely removed the need for the switch by assigning the compile time known hashes to the entries in the enum table. This works at compile time (even with no optimisations turned on): https://godbolt.org/z/4YYhrz7To

I concluded that the switch was too problematic to improve. I learnt that a function can't immediately proceed the `case` keyword which seems like a big limitation w.r.t. our aims.

Reply:
Yes! It's actually constexpr all the way down to the integer! Love it.